### PR TITLE
Event type is inferred from the target

### PR DESCRIPTION
**Pull request by Ayub-Begimkulov**

### CODE REVIEW

**Comment by Ayub-Begimkulov** [comment]
TODO: We need to test that all types work correctly with unions (type distribution) and correctly handle edge cases.

File: test/type-tests/binding.type-test.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  listener: Listener<UIEvent | BeforeUnloadEvent>;
+  options?: boolean | AddEventListenerOptions;
+}>();
+
+// correctly works with union target and type
+expectTypeOf<Binding<Window | Element, 'abort' | 'beforeunload'>>().toEqualTypeOf<{
+  type: 'abort' | 'beforeunload';
+  listener: Listener<UIEvent | Event | BeforeUnloadEvent>;
+  options?: boolean | AddEventListenerOptions;
+}>();
```
Comment (Ayub-Begimkulov) [inline test/type-tests/binding.type-test.ts:38]: tested potential edge cases, do you know any other unusual cases that we need to test?

File: src/types.ts [hunk 2]
```diff
@@ -1,7 +1,32 @@
...
+  ? `on${Type}` extends keyof Target
+    ? GetEventTypeFromListener<
+        // remove types that aren't assignable to `AnyFunction`
+        // so that we don't end up with union like `MouseEvent | Event`
+        Extract<Target[`on${Type}`], AnyFunction>
+      >
+    : Event
+  : never;
+
+type GetEventTypeFromListener<T extends AnyFunction> = T extends (this: any, event: infer U) => any
```
Comment (alexreardon) [inline src/types.ts:15]: technically an event listener can be a function or an object with a `handleEvent` property (my current types don't account for this either

File: test/type-tests/binding.type-test.ts (hunk 1 above)
Comment (alexreardon) [reply test/type-tests/binding.type-test.ts:38]: This looks good to me

File: src/bind-all.ts [hunk 3]
```diff
@@ -35,6 +35,85 @@ function getBinding(original: Binding, sharedOptions?: boolean | AddEventListene
...
+  Type6 extends string,
+>(
+  target: Target,
+  bindings: [
+    Binding<Target, Type1>,
+    Binding<Target, Type2>,
+    Binding<Target, Type3>,
+    Binding<Target, Type4>,
+    Binding<Target, Type5>,
+    Binding<Target, Type6>,
```
Comment (alexreardon) [inline src/bind-all.ts:108]: Will the typing stop working after 6 options? Is there a way to support any amount?

**Review by alexreardon — APPROVED** [review]

**Comment by alexreardon** [comment]
I will have a play with this one

**Comment by alexreardon** [comment]
This is truly great work @Ayub-Begimkulov

**Comment by alexreardon** [comment]
If you are happy with this @Ayub-Begimkulov, I can merge it and do a release

**Comment by alexreardon** [comment]
I thought of something: `this`.

For event listener functions, the `this` is set to the `EventTarget` that the listener is bound to

```ts
button.addEventListener('click', function onClick(event) {
	console.log(this) // button
});
```

For `EventListener` objects, the `this` is set to the `EventListener` object

```ts
const object = {
	handleEvent() {
		console.log(this) // object
	}
}
```

Could you please update the types to reflect the correct `this` bindings?

**Comment by alexreardon** [comment]
Here is the WIP 2.0.0 release notes. I would love to add something about the `this` typing as well

---


# New: full typescript support 🥳

Thanks to the incredible work of @Ayub-Begimkulov, `bind-event-listener` now has full typescript support. 

Event listener functions will now have the correct `event` type inferred based on the provided arguments.

```diff
  const button = document.querySelector('button');
  invariant(button instanceof HTMLElement);

  bind(button, {
    type: 'click',
+    // `event` is now correctly set as a `MouseEvent`
    listener: function onClick(event) {
+   // can safely use the `MouseEvent` properties
      console.log(event.button);
   },
  });
```

Incorrect event typings will now cause a typescript error

```diff
+ // This is now correctly an error as the `event` should be a `MouseEvent`
const onClick = function onClick(event: KeyboardEvent) {

}

bind(button, {
  type: 'click',
   listener: onClick,
});
```

## Breaking change

The improvements require TypeScript `4.1+` so this change has been listed as a breaking change. Also, because the types are getting more correct and a lot tighter, this is likely to cause some (correct) typescript breakages in projects. The number of these might be significant for larger projects. So on balance, I thought that a `major` correctly flags this release as higher risk as it is likely to 'break' a few consumers types.

File: src/types.ts [hunk 4]
```diff
@@ -1,7 +1,32 @@
...
+    : Event
+  : Event;
+
+export type Binding<Target extends EventTarget = EventTarget, Type extends string = string> = {
+  type: Type;
+  listener: Listener<GetEventType<Target, Type>>;
   options?: boolean | AddEventListenerOptions;
 };
+
+export type Listener<Ev extends Event> =
```
Comment (alexreardon) [inline src/types.ts:27]: Can we tighten this up so that `this` is set correctly?

**Review by alexreardon — CHANGES_REQUESTED** [review]

File: src/types.ts (hunk 4 above)
Comment (Ayub-Begimkulov) [reply src/types.ts:27]: Sure, we can.

File: src/bind-all.ts (hunk 3 above)
Comment (Ayub-Begimkulov) [reply src/bind-all.ts:108]: It won't stop working, but the event type will fall back to `Event`. This happens due to the reason that we need to know the `type` to resolve the event correctly. Therefore, we need to have a generic type for each element of the binding.   

I tried another way of doing it: I built a union of bindings from all `on*` properties, something like this:

```ts
| {
  type: 'abort',
  listener: Listener<UIEvent, Window>,
  ...
}
| ... 
| {
  type: 'mousemove',
  listener: Listener<MouseEvent, Window>,
  ...
}
```

And this implementation didn't require a function overload. However, if it wasn't working correctly with "unknown" events(e.g. `DOMContenetLoaded`). 

Because once I added a fallback case to the union (`{type: string, listener: Listener<Event, EventTareget>}`), the listener type started to be `any` for "known" events (e.g. 'mousemove'), since `type` was also always matching the fallback implementation.

So, right now I don't have any good idea about solving this problem, but I think that 6 handlers are more than enough. 
We could increase it up 8-10, but I don't think that there is any practical use case where you need more than this.

Comment (Ayub-Begimkulov) [reply src/bind-all.ts:108]: So for now I try to look at it as a necessary evil 😂

**Comment by alexreardon** [comment]
Epic work. I'll take a look asap

**Review by alexreardon — APPROVED** [review]
